crates/subspace-verification/src/lib.rs

```diff
@@ -19,6 +19,318 @@
 #![warn(rust_2018_idioms, missing_debug_implementations, missing_docs)]
 #![cfg_attr(not(feature = "std"), no_std)]
 
-mod verification;
+use core::mem;
+use schnorrkel::context::SigningContext;
+use schnorrkel::vrf::VRFOutput;
+use schnorrkel::{SignatureError, SignatureResult};
+use sp_arithmetic::traits::SaturatedConversion;
+use subspace_archiving::archiver;
+use subspace_core_primitives::{
+    crypto, PieceIndex, PieceIndexHash, Randomness, Salt, Sha256Hash, Solution, Tag, TagSignature,
+    RANDOMNESS_CONTEXT, RANDOMNESS_LENGTH, SALT_HASHING_PREFIX, SALT_SIZE, U256,
+};
+use subspace_solving::{
+    create_tag_signature_transcript, derive_global_challenge, derive_target, is_tag_valid,
+    verify_local_challenge, verify_tag_signature, SubspaceCodec,
+};
 
-pub use verification::*;
+const SALT_HASHING_PREFIX_LEN: usize = SALT_HASHING_PREFIX.len();
+
+/// Errors encountered by the Subspace consensus primitives.
+#[derive(Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum Error {
+    /// Tag verification failed
+    #[cfg_attr(feature = "thiserror", error("Invalid tag for salt"))]
+    InvalidTag,
+
+    /// Piece encoding is invalid
+    #[cfg_attr(feature = "thiserror", error("Invalid piece encoding"))]
+    InvalidPieceEncoding,
+
+    /// Piece verification failed
+    #[cfg_attr(feature = "thiserror", error("Invalid piece"))]
+    InvalidPiece,
+
+    /// Invalid Local challenge
+    #[cfg_attr(feature = "thiserror", error("Invalid local challenge"))]
+    InvalidLocalChallenge(SignatureError),
+
+    /// Solution is outside the challenge range
+    #[cfg_attr(feature = "thiserror", error("Solution is outside the solution range"))]
+    OutsideSolutionRange,
+
+    /// Invalid solution signature
+    #[cfg_attr(feature = "thiserror", error("Invalid solution signature"))]
+    InvalidSolutionSignature(SignatureError),
+
+    /// Solution is outside the MaxPlot
+    #[cfg_attr(feature = "thiserror", error("Solution is outside max plot"))]
+    OutsideMaxPlot,
+}
+
+/// Check the reward signature validity.
+pub fn check_reward_signature<FarmerPublicKey, RewardSignature>(
+    hash: &[u8],
+    signature: &RewardSignature,
+    public_key: &FarmerPublicKey,
+    reward_signing_context: &SigningContext,
+) -> Result<(), SignatureError>
+where
+    FarmerPublicKey: AsRef<[u8]>,
+    RewardSignature: AsRef<[u8]>,
```

Comment:
Well, whether it is `impl` or explicit generic, it is still a slice of memory of unknown length. Can you switch to concrete wrapper types from `subspace-core-primitives` instead?